### DISCUSSION TITLE
Mutation function arguments type correction

### DISCUSSION THREAD

**Discussion by Lisenish** [TanStack/query#1085] [General]
I wanted to do a PR but decided to discuss before that. 

I think we have a mistake with types of mutation function here:
https://github.com/tannerlinsley/react-query/blob/master/src/core/types.ts#L264
https://github.com/tannerlinsley/react-query/blob/master/src/core/types.ts#L273

I suppose it should be `...variables: TVariables` instead of just `variables: TVariable` because now if you have several arguments in mutation function there is no way to type them all in a clean way. 

E.g. I want to provide types for two arguments`TVariables = [string, number]`, but now if I do so it will lead to a situation where I have to call mutation function in that way `mutation(['someStr', 42])` and I want it to be `mutation('someStr', 42)`

**Comment by TkDodo** [comment]
I don't think this would work. As you have pointed out with your second link, there is an overloaded function that has the config as the last parameter. You can only have varargs at _the end_ of a parameter list, so, to take your example:

`mutation('someStr', 42, { onSuccess })`

wouldn't really work I think. TypeScript also yields the following error if I try to do that in a [playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAKgagQwE4EsECMA2EDOUC8UA2jsKgHYDmANFOQK4C26ESAugFAcDGA9uaSiN6wBMGiEAFADpZAN2RosuAFyxEqDNhy0+5AGYpKa3ugBWEbsACUBAHxQA3gF8gA)

`A rest parameter must be last in a parameter list.(1014)`

**Reply by Lisenish** [reply]
@TkDodo Actually it's a different type and not an overload but anyway, you were right I missed config object for the second type during my testing.

But thanks to TS 4.0 variadic tuples I think we can make it work, take a look at [your fixed example](https://www.typescriptlang.org/play?ssl=5&ssc=16&pln=5&pc=18#code/C4TwDgpgBAKgagQwE4EsECMA2EDOUC8UA2jsKgHYDmANFOQK4C26ESAugFAcDGA9uaSiN6wBMGiEAFADpZAN2RosuAFzFZ0+IozYctPuQBmKSmt7oAVhG7A2ASgIA+KAG8Avl2GjxkgOTjSX1oAFgAmWhd+AGV6bm5cHDVJB3xnAxxebGlMXko-AAkUWABpABFeABNeAEJfOzc7IA)

**Reply by TkDodo** [reply]
oh that is really nice :)

**Comment by stefanprobst** [comment]
will this be fixed in v3?
